File: python/aide_frame/http_server.py
"""
Simple HTTP server utilities for aide-frame applications.

Provides a lightweight HTTP server with JSON/HTML response helpers
and integrated docs/help route handling.

Usage:
    from aide_frame.http_server import HttpServer, JsonHandler

    class MyHandler(JsonHandler):
        def get(self, path, params):
            if path == '/': return self.file('index.html')
            if path == '/api/status': return {'ok': True}

        def post(self, path, data):
            if path == '/api/echo': return data

    server = HttpServer(port=8080, handler_class=MyHandler)
    server.run()  # Blocking, handles Ctrl+C
"""

import os
import json
import signal
import socket
import time
import threading
from http.server import HTTPServer, BaseHTTPRequestHandler
from urllib.parse import urlparse, parse_qs
from typing import Optional, Any, Dict, Type

from . import paths, http_routes, update_routes
from .log import logger


def get_server_url(port: int, platform: str = None) -> str:
    """
    Get the best URL to reach the server.

    Tries in order:
    1. For WSL2: localhost (Windows host can access via localhost)
    2. Local IP via UDP socket trick
    3. FQDN if available
    4. Hostname as fallback

    Args:
        port: Server port
        platform: Platform string (e.g., 'wsl2', 'raspi', 'linux')

    Returns:
        URL string like "http://192.168.1.100:8080"
    """
    hostname = socket.gethostname()

    # WSL2: localhost works for Windows host access
    if platform == 'wsl2':
        return f"http://localhost:{port}"

    # Try to get local IP via UDP socket (doesn't actually send anything)
    try:
        s = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
        s.connect(("8.8.8.8", 80))
        ip = s.getsockname()[0]
        s.close()
        return f"http://{ip}:{port}"
    except Exception:
        pass

    # Try FQDN
    try:
        fqdn = socket.getfqdn()
        if fqdn and fqdn != hostname and '.' in fqdn:
            return f"http://{fqdn}:{port}"
    except Exception:
        pass

    return f"http://{hostname}:{port}"


def restart_server(delay: float = 0.5):
    """
    Restart the server process after a short delay.

    This exits the process with code 0, expecting systemd or similar
    to restart it automatically.

    Args:
        delay: Seconds to wait before exit (allows response to be sent)

    Returns:
        Dict with success message (for JSON response)
    """
    def delayed_exit():
        time.sleep(delay)
        os._exit(0)
    threading.Thread(target=delayed_exit, daemon=True).start()
    return {"success": True, "message": "Restarting..."}


class JsonHandler(BaseHTTPRequestHandler):
    """
    HTTP request handler with JSON/HTML convenience methods.

    Subclass and override get() and post() methods.
    Return a dict for JSON, a string for HTML, or None if handled manually.
    """

    # Set by HttpServer
    app_dir: str = None
    static_dir: str = None
    docs_config: Optional[http_routes.DocsConfig] = None
    update_config: Optional[update_routes.UpdateConfig] = None

    def log_message(self, format, *args):
        """Suppress default logging, use our logger instead."""
        # args[0] is like "GET /path HTTP/1.1" - extract method and path
        # But for errors, args[0] can be an int (error code)
        if not args or not isinstance(args[0], str):
            return
        parts = args[0].split()
        if len(parts) >= 2:
            path = parts[1]
            # Skip static file requests (html, css, js, images)
            if path.startswith('/static/') or path.endswith(('.html', '.css', '.js', '.png', '.jpg', '.ico', '.svg', '.woff', '.woff2')):
                return
            # Only log API calls
            if path.startswith('/api/') or parts[0] == 'POST':
                logger.debug(f"HTTP: {parts[0]} {path}")

    # -------------------------------------------------------------------------
    # Response helpers
    # -------------------------------------------------------------------------

    def send_json(self, data: dict, status: int = 200):
        """Send JSON response."""
        self.send_response(status)
        self.send_header('Content-Type', 'application/json')
        self.send_header('Access-Control-Allow-Origin', '*')
        self.end_headers()
        body = json.dumps(data)
        # Log response preview (first 80 chars)
        preview = body[:80] + '...' if len(body) > 80 else body
        logger.debug(f"  -> {status} {preview}")
        self.wfile.write(body.encode())

    def send_html(self, content: str, status: int = 200):
        """Send HTML response."""
        self.send_response(status)
        self.send_header('Content-Type', 'text/html; charset=utf-8')
        self.end_headers()
        self.wfile.write(content.encode())

    def send_text(self, content: str, content_type: str = 'text/plain', status: int = 200):
        """Send text response with custom content type."""
        self.send_response(status)
        self.send_header('Content-Type', f'{content_type}; charset=utf-8')
        self.end_headers()
        self.wfile.write(content.encode())

    def send_file(self, filepath: str, content_type: str = None, binary: bool = False):
        """Send file content."""
        if not os.path.isfile(filepath):
            self.send_json({'error': 'File not found'}, 404)
            return

        if content_type is None:
            ext = os.path.splitext(filepath)[1].lower()
            content_type = paths.MIME_TYPES.get(ext, 'application/octet-stream')
            binary = ext in ('.png', '.jpg', '.jpeg', '.gif', '.ico', '.webp')

        try:
            mode = 'rb' if binary else 'r'
            encoding = None if binary else 'utf-8'
            with open(filepath, mode, encoding=encoding) as f:
                content = f.read()

            self.send_response(200)
            self.send_header('Content-Type', content_type)
            self.end_headers()

            if binary:
                self.wfile.write(content)
            else:
                self.wfile.write(content.encode())
        except Exception as e:
            self.send_json({'error': str(e)}, 500)

    def file(self, filename: str) -> None:
        """Serve a file from the app's static directory. Returns None to signal handled."""
        if self.static_dir:
            filepath = os.path.join(self.static_dir, filename)
            self.send_file(filepath)
        else:
            self.send_json({'error': 'Static directory not configured'}, 500)
        return None

    # -------------------------------------------------------------------------
    # Request handling
    # -------------------------------------------------------------------------

    def do_GET(self):
        """Handle GET requests."""
        parsed = urlparse(self.path)
        path = parsed.path

        # Let http_routes handle docs/help routes first
        if self.docs_config and http_routes.handle_request(self, self.path, self.docs_config):
            return

        # Let update_routes handle update routes
        if self.update_config and update_routes.handle_update_request(self, path, 'GET', {}, self.update_config):
            return

        # Serve static files from /static/ path (e.g., locales, images)
        if path.startswith('/static/') and self.static_dir:
            filename = path[8:]  # Remove '/static/' prefix
            if '..' not in filename and not filename.startswith('/'):
                filepath = os.path.join(self.static_dir, filename)
                if os.path.isfile(filepath):
                    self.send_file(filepath)
                    return

        params = {k: v[0] if len(v) == 1 else v for k, v in parse_qs(parsed.query).items()}

        result = self.get(path, params)
        self._send_result(result)

    def do_POST(self):
        """Handle POST requests."""
        parsed = urlparse(self.path)
        path = parsed.path

        # Read and parse body
        content_length = int(self.headers.get('Content-Length', 0))
        body = self.rfile.read(content_length).decode('utf-8') if content_length > 0 else '{}'

        try:
            data = json.loads(body) if body else {}
        except json.JSONDecodeError:
            self.send_json({'error': 'Invalid JSON'}, 400)
            return

        # Let update_routes handle update routes
        if self.update_config and update_routes.handle_update_request(self, path, 'POST', data, self.update_config):
            return

        result = self.post(path, data)
        self._send_result(result)

    def _send_result(self, result):
        """Send result based on type."""
        if result is None:
            return  # Already handled
        elif isinstance(result, dict):
            self.send_json(result)
        elif isinstance(result, str):
            self.send_html(result)
        elif isinstance(result, tuple) and len(result) == 2:
            # (data, status) tuple
            data, status = result
            if isinstance(data, dict):
                self.send_json(data, status)
            else:
                self.send_html(str(data), status)

    # -------------------------------------------------------------------------
    # Override these in subclass
    # -------------------------------------------------------------------------

    def get(self, path: str, params: Dict[str, Any]) -> Optional[Any]:
        """
        Handle GET request. Override in subclass.

        Args:
            path: URL path (e.g., '/api/status')
            params: Query parameters as dict

        Returns:
            - dict: Sent as JSON
            - str: Sent as HTML
            - None: Already handled (e.g., via self.file())
            - (data, status): Tuple with status code
        """
        return {'error': 'Not found'}, 404

    def post(self, path: str, data: Dict[str, Any]) -> Optional[Any]:
        """
        Handle POST request. Override in subclass.

        Args:
            path: URL path
            data: Parsed JSON body

        Returns:
            Same as get()
        """
        return {'error': 'Not found'}, 404


class HttpServer:
    """
    Simple HTTP server with lifecycle management.

    Usage:
        server = HttpServer(port=8080, handler_class=MyHandler)
        server.run()  # Blocking

        # Or non-blocking:
        server.start()
        # ... do other things ...
        server.stop()
    """

    def __init__(
        self,
        port: int = 8080,
        handler_class: Type[JsonHandler] = JsonHandler,
        app_dir: str = None,
        static_dir: str = None,
        docs_config: http_routes.DocsConfig = None,
        update_config: update_routes.UpdateConfig = None,
    ):
        """
        Initialize HTTP server.

        Args:
            port: Port to listen on
            handler_class: Handler class (subclass of JsonHandler)
            app_dir: Application directory (for paths.init)
            static_dir: Directory for static files (default: app_dir/static)
            docs_config: DocsConfig for docs/help routes
            update_config: UpdateConfig for remote update functionality
        """
        self.port = port
        self.handler_class = handler_class
        self.app_dir = app_dir
        self.static_dir = static_dir or (os.path.join(app_dir, 'static') if app_dir else None)
        self.docs_config = docs_config
        self.update_config = update_config

        self._server: Optional[HTTPServer] = None
        self._thread: Optional[threading.Thread] = None
        self._running = False

    def start(self):
        """Start the server in a background thread."""
        if self._running:
            return

        # Initialize paths if app_dir provided
        if self.app_dir:
            paths.init(self.app_dir)

        # Configure handler class
        self.handler_class.app_dir = self.app_dir
        self.handler_class.static_dir = self.static_dir
        self.handler_class.docs_config = self.docs_config
        self.handler_class.update_config = self.update_config

        self._server = HTTPServer(('0.0.0.0', self.port), self.handler_class)
        self._thread = threading.Thread(target=self._server.serve_forever, daemon=True)
        self._thread.start()
        self._running = True

        logger.info(f"Server started on http://localhost:{self.port}")

    def stop(self):
        """Stop the server."""
        if self._server:
            self._server.shutdown()
            self._running = False
            logger.info("Server stopped")

    def run(self):
        """Start server and block until Ctrl+C."""
        self.start()

        # Handle Ctrl+C gracefully
        shutdown_event = threading.Event()

        def signal_handler(sig, frame):
            logger.info("Shutting down...")
            shutdown_event.set()

        signal.signal(signal.SIGINT, signal_handler)
        signal.signal(signal.SIGTERM, signal_handler)

        logger.info("Press Ctrl+C to stop")

        # Wait for shutdown signal
        try:
            shutdown_event.wait()
        except KeyboardInterrupt:
            pass
        finally:
            self.stop()


__all__ = [
    'JsonHandler',
    'HttpServer',
    'get_server_url',
    'restart_server',
]

File: js/aide_frame/src/http-server.js
/**
 * Simple HTTP server utilities for aide-frame applications.
 *
 * Provides a lightweight HTTP server based on Express with JSON/HTML response helpers
 * and integrated docs/help route handling.
 *
 * Usage:
 *   const { HttpServer } = require('aide-frame');
 *
 *   const server = new HttpServer({
 *       port: 8080,
 *       appDir: __dirname,
 *       docsConfig: { appName: 'My App' }
 *   });
 *
 *   server.addRoutes((app) => {
 *       app.get('/', (req, res) => res.sendFile('index.html'));
 *       app.get('/api/status', (req, res) => res.json({ ok: true }));
 *   });
 *
 *   server.run();  // Blocking, handles Ctrl+C
 */

const express = require('express');
const http = require('http');
const https = require('https');
const path = require('path');
const fs = require('fs');
const os = require('os');
const dgram = require('dgram');

const paths = require('./paths');
const { logger } = require('./log');

/**
 * Get the best URL to reach the server.
 *
 * Tries in order:
 * 1. For WSL2: localhost (Windows host can access via localhost)
 * 2. Local IP via network interfaces
 * 3. Hostname as fallback
 *
 * @param {number} port - Server port
 * @param {string|null} platform - Platform string (e.g., 'wsl2', 'raspi', 'linux')
 * @returns {string} URL string like "http://192.168.1.100:8080"
 */
function getServerUrl(port, platform = null, basePath = '') {
    const hostname = os.hostname();
    const suffix = basePath ? basePath + '/' : '';

    // WSL2: localhost works for Windows host access
    if (platform === 'wsl2') {
        return `http://localhost:${port}${suffix}`;
    }

    // Try to get local IP from network interfaces
    try {
        const interfaces = os.networkInterfaces();
        for (const name of Object.keys(interfaces)) {
            for (const iface of interfaces[name]) {
                // Skip internal (loopback) and non-IPv4 addresses
                if (!iface.internal && iface.family === 'IPv4') {
                    return `http://${iface.address}:${port}${suffix}`;
                }
            }
        }
    } catch (e) {
        // Ignore errors
    }

    return `http://${hostname}:${port}${suffix}`;
}

/**
 * Restart the server process after a short delay.
 *
 * This exits the process with code 0, expecting systemd or similar
 * to restart it automatically.
 *
 * @param {number} delay - Seconds to wait before exit (allows response to be sent)
 * @returns {object} Object with success message (for JSON response)
 */
function restartServer(delay = 0.5) {
    setTimeout(() => {
        process.exit(0);
    }, delay * 1000);
    return { success: true, message: 'Restarting...' };
}

/**
 * HTTP Server with lifecycle management.
 */
class HttpServer {
    /**
     * Initialize HTTP server.
     *
     * @param {object} options - Configuration options
     * @param {number} options.port - Port to listen on (default: 8080)
     * @param {string} options.appDir - Application directory (for paths.init)
     * @param {string} options.staticDir - Directory for static files (default: appDir/static)
     * @param {object} options.docsConfig - DocsConfig for docs/help routes
     * @param {object} options.updateConfig - UpdateConfig for remote update functionality
     * @param {string} [options.httpsCertPath] - Path to TLS certificate (PEM). When set
     *   together with httpsKeyPath, the server speaks HTTPS instead of HTTP.
     * @param {string} [options.httpsKeyPath]  - Path to TLS private key (PEM).
     */
    constructor(options = {}) {
        this.port = options.port || 8080;
        this.basePath = options.basePath || '';
        this.appDir = options.appDir;
        this.staticDir = options.staticDir || (options.appDir ? path.join(options.appDir, 'static') : null);
        this.docsConfig = options.docsConfig || null;
        this.updateConfig = options.updateConfig || null;
        this.httpsCertPath = options.httpsCertPath || null;
        this.httpsKeyPath  = options.httpsKeyPath  || null;
        if (!!this.httpsCertPath !== !!this.httpsKeyPath) {
            throw new Error('HttpServer: httpsCertPath and httpsKeyPath must both be set, or both be omitted.');
        }

        this._server = null;
        this._running = false;
        this._scheme  = (this.httpsCertPath && this.httpsKeyPath) ? 'https' : 'http';

        // Initialize Express app
        this.app = express();
        this.app.use(express.json({ limit: '5mb' }));

        // Register early middleware (e.g., cookie parser) before any routes
        if (options.earlyMiddleware) {
            for (const mw of options.earlyMiddleware) {
                this.app.use(mw);
            }
        }

        // Initialize paths if appDir provided
        if (this.appDir) {
            paths.init(this.appDir);
        }

        this._setupRoutes();
    }

    /**
     * Setup default routes for static files and framework features.
     * @private
     */
    _setupRoutes() {
        // Framework static files (/static/frame/*)
        const aideFrameStaticDir = paths.get('AIDE_FRAME_STATIC_DIR');
        if (aideFrameStaticDir && fs.existsSync(aideFrameStaticDir)) {
            this.app.use('/static/frame', express.static(aideFrameStaticDir));
        }

        // App static files (/static/*)
        if (this.staticDir && fs.existsSync(this.staticDir)) {
            this.app.use('/static', express.static(this.staticDir));
        }

        // Register docs/help routes if docsConfig provided
        if (this.docsConfig) {
            const httpRoutes = require('./http-routes');
            httpRoutes.register(this.app, { ...this.docsConfig, basePath: this.basePath });
        }

        // Register update routes if updateConfig provided
        if (this.updateConfig) {
            const updateRoutes = require('./update-routes');
            updateRoutes.register(this.app, { ...this.updateConfig, basePath: this.basePath });
        }
    }

    /**
     * Add custom routes to the Express app.
     *
     * @param {function} callback - Function that receives the Express app
     */
    addRoutes(callback) {
        callback(this.app);
    }

    /**
     * Get the Express app instance for advanced configuration.
     * @returns {express.Application}
     */
    getApp() {
        return this.app;
    }

    /**
     * Start the server.
     * @returns {Promise} Resolves when server is listening
     */
    start() {
        return new Promise((resolve, reject) => {
            if (this._running) {
                resolve();
                return;
            }

            let listenApp = this.app;

            if (this.basePath) {
                // Mount entire app as sub-app under basePath
                const wrapper = express();
                wrapper.enable('strict routing');
                wrapper.get(this.basePath, (req, res) => res.redirect(301, this.basePath + '/'));
                wrapper.use(this.basePath, this.app);
                listenApp = wrapper;
            }

            // A failing listen() must REJECT this promise. Without an 'error' handler the
            // `listening` callback simply never fires, so the promise neither resolves nor
            // rejects — it dangles, the caller's .then() never runs, and the emitted 'error'
            // escapes as an uncaughtException. A logger that catches uncaughtException (as
            // aide-rap's winston does) then keeps the process ALIVE: fully booted, holding
            // the system's server.lock, reported "online" by its supervisor — and serving
            // nothing. It also blocks the next start, which is then refused with the
            // singleton-lock message, so the symptom presents as an entirely different bug.
            // Lehrgeld 2026-07-20 (study@corno): five failed starts in a row, each leaving
            // the zombie that broke the next.
            //
            // A dead listener is unrecoverable, so the caller must be able to fail fast and
            // exit non-zero. The message names port and cause: an operator who reads
            // "EADDRINUSE :18349" is done in seconds.
            const onListenError = (err) => {
                const why = err && err.code === 'EADDRINUSE'
                    ? `port ${this.port} is already in use (EADDRINUSE) — another instance is probably running`
                    : `${err && err.code ? err.code + ': ' : ''}${err && err.message ? err.message : String(err)}`;
                const e = new Error(`HttpServer: cannot listen on ${this._scheme}://0.0.0.0:${this.port} — ${why}`);
                e.code = err && err.code;
                e.cause = err;
                this._running = false;
                reject(e);
            };

            if (this._scheme === 'https') {
                const cert = fs.readFileSync(this.httpsCertPath);
                const key  = fs.readFileSync(this.httpsKeyPath);
                this._server = https.createServer({ cert, key }, listenApp).listen(this.port, '0.0.0.0', () => {
                    this._running = true;
                    this._url = `https://localhost:${this.port}${this.basePath || ''}`;
                    resolve();
                });
            } else {
                this._server = listenApp.listen(this.port, '0.0.0.0', () => {
                    this._running = true;
                    this._url = `http://localhost:${this.port}${this.basePath || ''}`;
                    resolve();
                });
            }
            // Bound AFTER listen() so it is attached to the actual server object; 'error' is
            // emitted asynchronously, so this still lands before any bind failure surfaces.
            this._server.on('error', onListenError);
        });
    }

    /**
     * Stop the server.
     */
    stop() {
        if (this._server) {
            this._server.close();
            this._running = false;
            logger.info('Server stopped');
        }
    }

    /**
     * Log the server URL. Call this after all startup tasks are complete.
     */
    logReady() {
        console.log('');
        logger.info(`Press Ctrl+C to stop`);
        logger.info(`Server started on ${this._url}`);
    }

    /**
     * Start server and block until Ctrl+C.
     */
    run() {
        // Handle Ctrl+C gracefully
        const shutdown = () => {
            logger.info('Shutting down...');
            this.stop();
            process.exit(0);
        };

        process.on('SIGINT', shutdown);
        process.on('SIGTERM', shutdown);

        this.start().then(() => this.logReady());
    }
}

module.exports = {
    HttpServer,
    getServerUrl,
    restartServer,
};
